File: dsa-problems/Trees/solutions/cpp/TRE-009-campus-vertical-order-weight.cpp
#include <iostream>
#include <vector>
#include <string>
#include <sstream>
#include <unordered_map>
#include <algorithm>
#include <queue>

using namespace std;

struct NodeInfo {
    int val;
    int weight;
    int depth;
};

class Solution {
public:
    vector<vector<int>> verticalOrderWithWeights(int n, const vector<int>& values,
                                                 const vector<int>& weights, const vector<int>& left,
                                                 const vector<int>& right, int W) {
        if (n == 0) return {};

        vector<bool> hasParent(n, false);
        for (int i = 0; i < n; i++) {
            if (left[i] != -1) hasParent[left[i]] = true;
            if (right[i] != -1) hasParent[right[i]] = true;
        }
        int root = 0;
        for (int i = 0; i < n; i++) {
            if (!hasParent[i]) {
                root = i;
                break;
            }
        }

        unordered_map<int, vector<NodeInfo>> cols;
        queue<pair<int, pair<int, int>>> q;
        vector<bool> visited(n, false);
        q.push({root, {0, 0}});
        visited[root] = true;

        int minCol = 0;
        int maxCol = 0;

        while (!q.empty()) {
            auto curr = q.front();
            q.pop();
            int u = curr.first;
            int c = curr.second.first;
            int d = curr.second.second;

            cols[c].push_back({values[u], weights[u], d});
            minCol = min(minCol, c);
            maxCol = max(maxCol, c);

            if (left[u] != -1 && !visited[left[u]]) {
                visited[left[u]] = true;
                q.push({left[u], {c - 1, d + 1}});
            }
            if (right[u] != -1 && !visited[right[u]]) {
                visited[right[u]] = true;
                q.push({right[u], {c + 1, d + 1}});
            }
        }

        vector<vector<int>> result;
        for (int c = minCol; c <= maxCol; c++) {
            auto it = cols.find(c);
            if (it == cols.end()) continue;
            auto& list = it->second;
            long long totalWeight = 0;
            for (const auto& node : list) totalWeight += node.weight;
            if (totalWeight >= W) {
                sort(list.begin(), list.end(), [](const NodeInfo& a, const NodeInfo& b) {
                    if (a.depth != b.depth) return a.depth < b.depth;
                    if (a.weight != b.weight) return a.weight > b.weight;
                    return a.val < b.val;
                });
                vector<int> colValues;
                for (const auto& node : list) colValues.push_back(node.val);
                result.push_back(colValues);
            }
        }
        return result;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<string> lines;
    string line;
    while (getline(cin, line)) {
        if (!line.empty()) {
            bool allSpace = true;
            for (char ch : line) {
                if (ch > ' ') {
                    allSpace = false;
                    break;
                }
            }
            if (!allSpace) lines.push_back(line);
        }
    }
    if (lines.empty()) return 0;

    int n = stoi(lines[0]);
    vector<int> values(n, 0), weights(n, 1), left(n, -1), right(n, -1);

    for (int i = 0; i < n && i + 1 < (int)lines.size(); i++) {
        stringstream ss(lines[i + 1]);
        vector<long long> parts;
        long long x;
        while (ss >> x) parts.push_back(x);
        if (parts.size() < 3) continue;
        values[i] = (int)parts[0];
        if (parts.size() >= 4) {
            weights[i] = (int)parts[1];
            left[i] = (int)parts[2];
            right[i] = (int)parts[3];
        } else {
            weights[i] = 1;
            left[i] = (int)parts[1];
            right[i] = (int)parts[2];
        }
    }

    int W = 0;
    if ((int)lines.size() > n + 1) {
        W = stoi(lines[n + 1]);
    }

    Solution solution;
    vector<vector<int>> cols = solution.verticalOrderWithWeights(n, values, weights, left, right, W);
    if (cols.empty()) {
        cout << "\n";
    } else {
        for (int i = 0; i < (int)cols.size(); i++) {
            for (int j = 0; j < (int)cols[i].size(); j++) {
                if (j) cout << ' ';
                cout << cols[i][j];
            }
            if (i + 1 < (int)cols.size()) cout << "\n";
        }
    }
    return 0;
}

File: dsa-problems/Trees/solutions/python/TRE-009-campus-vertical-order-weight.py
import sys
sys.setrecursionlimit(200000)
from collections import deque, defaultdict

def vertical_order_with_weights(n: int, values: list[int], weights: list[int],
                                left: list[int], right: list[int], W: int) -> list[list[int]]:
    if n == 0:
        return []

    # Identify Root
    has_parent = [False] * n
    for i in range(n):
        if left[i] != -1: has_parent[left[i]] = True
        if right[i] != -1: has_parent[right[i]] = True
        
    root = 0
    for i in range(n):
        if not has_parent[i]:
            root = i
            break
            
    cols = defaultdict(list)
    q = deque([(root, 0, 0)]) # u, col, depth
    
    min_col, max_col = 0, 0
    
    visited = {root} # Safety against cycles/bad input
    
    while q:
        u, c, d = q.popleft()
        cols[c].append((values[u], weights[u], d))
        min_col = min(min_col, c)
        max_col = max(max_col, c)
        
        if left[u] != -1 and left[u] not in visited:
            visited.add(left[u])
            q.append((left[u], c - 1, d + 1))
        if right[u] != -1 and right[u] not in visited:
            visited.add(right[u])
            q.append((right[u], c + 1, d + 1))
            
    result = []
    for c in range(min_col, max_col + 1):
        if c in cols:
            nodes = cols[c]
            total_weight = sum(node[1] for node in nodes)
            if total_weight >= W:
                # Priority: depth asc (x[2]), weight desc (-x[1]), value asc (x[0])
                nodes.sort(key=lambda x: (x[2], -x[1], x[0]))
                result.append([x[0] for x in nodes])
                
    return result

def main():
    lines = sys.stdin.read().splitlines()
    if not lines: return
    valid_lines = [l.strip() for l in lines if l.strip()]
    if not valid_lines: return
    iterator = iter(valid_lines)
    
    try:
        n = int(next(iterator))
        values = [0] * n
        weights = [0] * n
        left = [0] * n
        right = [0] * n
        
        for i in range(n):
            line = next(iterator)
            parts = list(map(int, line.split()))
            # 4 tokens: val weight left right
            # 3 tokens: val left right (default weight 1)
            
            values[i] = parts[0]
            if len(parts) >= 4:
                weights[i] = parts[1]
                left[i] = parts[2]
                right[i] = parts[3]
            else:
                weights[i] = 1
                left[i] = parts[1]
                right[i] = parts[2]
                
        # W might be on next line
        try:
            line = next(iterator)
            W = int(line)
        except StopIteration:
            W = 0
            
        cols = vertical_order_with_weights(n, values, weights, left, right, W)
        if not cols:
            print()
        else:
            print("\n".join(" ".join(str(x) for x in col) for col in cols))
            
    except StopIteration:
        pass

if __name__ == "__main__":
    main()
